# 1: Min-Max Нормализация

## **Задание**

Напишите программу, реализующую процесс нормализации входных данных методом **Min-Max**. Затем выполните сравнение результатов ручной реализации алгоритма с функцией из библиотеки `scikit-learn`. Программа должна выводить оба результата нормализации.

### Входные данные:
Набор данных представлен следующим двумерным массивом NumPy:

$$
\begin{bmatrix}
-1 & 2 \\
-0.5 & 6 \\
0 & 10 \\
1 & 18
\end{bmatrix}
$$

### Требования к программе:
1. Напишите функцию `manual_min_max`, которая принимает массив данных и возвращает матрицу с элементами, приведенными к диапазону $[0, 1]$ по каждому столбцу отдельно.
   
   Формула для расчета Min-Max нормализации по одному столбцу выглядит так:
   
$$
x_{norm} = \frac{x - x_{\text{min}}}{x_{\text{max}} - x_{\text{min}}}
$$

   Где:
   - $x$ — значение элемента,
   - $x_{\text{min}}$ — минимальное значение в данном столбце,
   - $x_{\text{max}}$ — максимальное значение в данном столбце.

2. Выполните самостоятельную реализацию, а также проведите аналогичную операцию с использованием встроенного класса `MinMaxScaler` из библиотеки `scikit-learn`.

3. Выведите обе нормализованные матрицы на экран и убедитесь, что ваши результаты совпадают с результатами стандартной реализации.

### Ожидаемый вывод программы:
Программа должна вывести две таблицы нормализованных значений: одну, рассчитанную вами вручную, и вторую, полученную с помощью библиотеки `scikit-learn`.

Пример вывода:

```
Ручная нормализация:
[[0.         0.        ]
 [0.33333333 0.28571429]
 [0.66666667 0.57142857]
 [1.         1.        ]]

MinMaxScaler (scikit-learn):
[[0.         0.        ]
 [0.33333333 0.28571429]
 [0.66666667 0.57142857]
 [1.         1.        ]]
```

Убедитесь, что ваша реализация совпадает с результатом работы встроенной функции.


Задание:

In [5]:
import numpy as np
from sklearn.preprocessing import MinMaxScaler

# Исходные данные
data = np.array([[-1, 2], [-0.5, 6], [0, 10], [1, 18]])

def manual_min_max(data):
    # Нормализация вручную через явный цикл по столбцам
    col_min = data.min(axis=0)
    col_max = data.max(axis=0)
    normalized = np.zeros_like(data, dtype=float)
    for j in range(data.shape[1]):
        for i in range(data.shape[0]):
            normalized[i, j] = (data[i, j] - col_min[j]) / (col_max[j] - col_min[j])
    return normalized

# Ручная нормализация
manual_result = manual_min_max(data)
print('Ручная нормализация:\n', manual_result)

# Проверка с помощью scikit-learn
scaler = MinMaxScaler()
sklearn_result = scaler.fit_transform(data)
print('\nMinMaxScaler (scikit-learn):\n', sklearn_result)

print('\nОбе матрицы идентичны, что говорит о одинаковом поведении встроенной функции и нашей реализации')


Ручная нормализация:
 [[0.   0.  ]
 [0.25 0.25]
 [0.5  0.5 ]
 [1.   1.  ]]

MinMaxScaler (scikit-learn):
 [[0.   0.  ]
 [0.25 0.25]
 [0.5  0.5 ]
 [1.   1.  ]]

Обе матрицы идентичны, что говорит о одинаковом поведении встроенной функции и нашей реализации


--------------------------------------------------------------------

# 2: Z-нормализация (стандартизация)

## Задание

Реализуйте алгоритм Z-нормализации (или стандартного отклонения) вручную и сравните полученный результат с реализацией из библиотеки Scikit-Learn (`StandardScaler`).

### Условия:
Вам дан следующий набор данных:

$$
\begin{bmatrix}
-1 & 2 \\
-0.5 & 6 \\
0 & 10 \\
1 & 18
\end{bmatrix}
$$

Вашей задачей является выполнение следующей последовательности действий:

1. Самостоятельно напишите функцию `manual_z_normalization()`, которая реализует формулу Z-нормализации по следующему принципу:

$$
x' = \frac{(x - \mu)}{\sigma},
$$

где:
- $x'$ — нормализованное значение,
- $\mu$ — среднее арифметическое элементов соответствующего столбца,
- $\sigma$ — стандартное отклонение элементов соответствующего столбца.

Формулы вычисления среднего ($\mu$) и стандартного отклонения ($\sigma$) по столбцам:

$$
\mu_j = \frac{\sum_{i=1}^{N}x_{ij}}{N},\quad\sigma_j=\sqrt{\frac{\sum_{i=1}^{N}(x_{ij}-\mu_j)^2}{N}},
$$

где $j$ обозначает номер столбца, $i$ — строку, $N$ — количество строк.

2. После самостоятельной реализации примените библиотеку Scikit-Learn (`StandardScaler`) для сравнения результатов.

3. Отобразите результаты обеих нормализаций и убедитесь, что они совпадают.

### Требуемый вывод:
Вы получите две нормализованные матрицы: одна, созданная вручную, вторая — с помощью `StandardScaler`. Приведите результаты вашего кода и сделайте вывод о совпадении результатов двух подходов.

Пример ожидаемого вывода:

```
Z-нормализация ручная:
[[-1.22474487 -1.34164079]
 [-0.40824829 -0.4472136 ]
 [ 0.40824829  0.4472136 ]
 [ 1.22474487  1.34164079]]

StandardScaler (scikit-learn):
[[-1.22474487 -1.34164079]
 [-0.40824829 -0.4472136 ]
 [ 0.40824829  0.4472136 ]
 [ 1.22474487  1.34164079]]
```

Таким образом, ваше решение должно подтвердить правильность реализации Z-нормализации и показать согласованность результатов.

Задание:

In [6]:
from sklearn.preprocessing import StandardScaler

# Данные остаются теми же
data = np.array([[-1, 2], [-0.5, 6], [0, 10], [1, 18]])

def manual_z_normalization(data):
    n = data.shape[0]
    # Среднее: сумма / N по каждому столбцу
    col_mean = data.sum(axis=0) / n
    # Стандартное отклонение по формуле из задания (ddof=0)
    col_std = np.sqrt(((data - col_mean) ** 2).sum(axis=0) / n)
    return (data - col_mean) / col_std

# Ручная нормализация
z_norm_manual = manual_z_normalization(data)
print('Z-нормализация ручная:\n', z_norm_manual)

# Использование scikit-learn
scaler = StandardScaler()
z_norm_sklearn = scaler.fit_transform(data)
print('\nStandardScaler (scikit-learn):\n', z_norm_sklearn)

# Вывод о совпадении результатов
if np.allclose(z_norm_manual, z_norm_sklearn):
  print("\nВывод: результаты совпадают — ручная реализация корректна.")
else:
  print("\nВывод: результаты расходятся — в реализации есть ошибка.")

Z-нормализация ручная:
 [[-1.18321596 -1.18321596]
 [-0.50709255 -0.50709255]
 [ 0.16903085  0.16903085]
 [ 1.52127766  1.52127766]]

StandardScaler (scikit-learn):
 [[-1.18321596 -1.18321596]
 [-0.50709255 -0.50709255]
 [ 0.16903085  0.16903085]
 [ 1.52127766  1.52127766]]

Вывод: результаты совпадают — ручная реализация корректна.


---------------------------------------------------------------------

# 3: Робастная нормализация

### Задание

**Цель:** Изучить работу метода `RobustScaler` библиотеки scikit-learn и применить его для нормализации набора данных с наличием выбросов.

#### Шаги выполнения задания:

1. **Теоретическая подготовка**
   - Ознакомьтесь с понятием масштабирования признаков (`feature scaling`) и различными методами нормализации данных.
   - Изучите метод робастной нормализации (`Robust Scaler`), включая принцип его работы и отличия от других методов нормализации.

2. **Подготовка среды**
   - Установите библиотеку scikit-learn, если она ещё не установлена, используя команду `pip install scikit-learn`.

3. **Создание тестового набора данных**
   - Используйте предложенный код для создания массива данных с выбросами. Массив состоит из четырёх пар значений, две из которых являются аномальными («выбросами»).

4. **Применение RobustScaler**
   - Примените класс `RobustScaler`, создав экземпляр класса и вызвав методы `.fit()` и `.transform()`. Это позволит нормализовать исходные данные с учётом наличия выбросов.

5. **Анализ результатов**
   - Выведите полученные нормализованные данные и сравните их с исходными значениями.
   - Объясните, почему выбран именно этот метод нормализации и каким образом он справляется с присутствием выбросов.

6. **Дополнительные исследования**
   - Попробуйте сравнить результаты нормализации с использованием других методов, таких как MinMaxScaler или StandardScaler, и сделайте выводы о преимуществах и недостатках каждого подхода.


In [9]:
from sklearn.preprocessing import RobustScaler

# Создаем новый массив с несколькими аномально большими значениями ("выбросы")
outlier_data = np.array([[1, 2], [2, 4], [3, 6], [100, 200]])  # Последние два значения — выбросы

print('Исходные данные:\n', outlier_data)

# Робастная нормализация
robust_scaler = RobustScaler()
robust_scaled_data = robust_scaler.fit_transform(outlier_data)
print('\nРобастная нормализация (RobustScaler):\n', robust_scaled_data)

# Сравнение с другими методами
mm_scaled = MinMaxScaler().fit_transform(outlier_data)
std_scaled = StandardScaler().fit_transform(outlier_data)
print('\nMin-Max (для сравнения):\n', mm_scaled)
print('\nStandardScaler (для сравнения):\n', std_scaled)

print('\n--- Анализ ---')
print('RobustScaler выбран потому, что данные содержат выбросы.')
print('Метод использует медиану и межквартильный размах (IQR = Q3 - Q1),')
print('которые нечувствительны к экстремальным значениям.')
print()
print('Сравнение: выброс сильно искажает Min-Max (все нормальные')
print('значения сжимаются к 0) и StandardScaler (смещает среднее и std).')
print('RobustScaler масштабирует нормальные строки в разумном диапазоне,')
print('а выброс просто получает большое значение, не искажая остальные.')


Исходные данные:
 [[  1   2]
 [  2   4]
 [  3   6]
 [100 200]]

Робастная нормализация (RobustScaler):
 [[-0.05882353 -0.05882353]
 [-0.01960784 -0.01960784]
 [ 0.01960784  0.01960784]
 [ 3.82352941  3.82352941]]

Min-Max (для сравнения):
 [[0.         0.        ]
 [0.01010101 0.01010101]
 [0.02020202 0.02020202]
 [1.         1.        ]]

StandardScaler (для сравнения):
 [[-0.60083218 -0.60083218]
 [-0.57727013 -0.57727013]
 [-0.55370809 -0.55370809]
 [ 1.7318104   1.7318104 ]]

--- Анализ ---
RobustScaler выбран потому, что данные содержат выбросы.
Метод использует медиану и межквартильный размах (IQR = Q3 - Q1),
которые нечувствительны к экстремальным значениям.

Сравнение: выброс сильно искажает Min-Max (все нормальные
значения сжимаются к 0) и StandardScaler (смещает среднее и std).
RobustScaler масштабирует нормальные строки в разумном диапазоне,
а выброс просто получает большое значение, не искажая остальные.


--------------------------------------------------

# 4: Применение различных методов нормализации на одном датасете

### Задание

**Цель:** Исследование различных методов нормализации данных на примере классического датасета "Iris" с применением трёх основных техник нормализации: Min-Max normalization, Z-normalization (StandardScaler), RobustScaler.

#### Шаги выполнения задания:

1. **Теоретическая подготовка**
   - Изучите понятие нормализации данных и её важность в машинном обучении.
   - Ознакомьтесь с тремя основными видами нормализации: Min-Max, Z-нормализацией и RobustScaler.
     - **Min-Max**: Приведение всех признаков к диапазону $[0, 1]$ путём линейного преобразования.
     - **Z-нормализация (StandardScaler)**: Преобразование данных таким образом, чтобы среднее значение стало равным нулю, а стандартное отклонение — единице.
     - **RobustScaler**: Метод, устойчивый к выбросам, основанный на медианах и межквартильных размахах.

2. **Загрузка данных**
   - Используем встроенный датасет "Iris" из библиотеки Scikit-Learn.
   - Создайте Pandas DataFrame с признаками цветка (длина и ширина чашелистика и лепестка).

3. **Нормализация данных**
   - Реализуйте три вида нормализации с помощью соответствующих классов из модуля `sklearn.preprocessing`: `MinMaxScaler`, `StandardScaler`, `RobustScaler`.
   - Проверьте правильность настройки и применения каждого метода путем трансформации исходных данных.

4. **Анализ результатов**
   - Посмотрите на первые пять строк каждой нормализованной таблицы.
   - Сделайте вывод о различиях между результатами применённых методов нормализации.
   - Обсудите особенности каждого метода, учитывая наличие возможных выбросов в реальных наборах данных.

5. **Заключение**
   - Подведите итог, какой метод наиболее подходит для вашего случая, исходя из особенностей ваших данных и требований конкретной модели.


In [11]:
import pandas as pd
from sklearn.datasets import load_iris

# Загрузка датасета Iris
X, _ = load_iris(return_X_y=True)
df = pd.DataFrame(X, columns=['sepal_length', 'sepal_width', 'petal_length', 'petal_width'])

# Применение трёх методов нормализации
scaled_mm= MinMaxScaler().fit_transform(df)
scaled_std = StandardScaler().fit_transform(df)
scaled_rb= RobustScaler().fit_transform(df)

print('Min-Max Normalized Data (первые 5 строк):')
print(scaled_mm[:5])

print('\nZ-Normalized Data (первые 5 строк):')
print(scaled_std[:5])

print('\nRobust Scaled Data (первые 5 строк):')
print(scaled_rb[:5])

# Проверка правильности применения каждого метода
print("\n--- Проверка корректности трансформации ---")

# Min-Max: все значения должны быть в диапазоне [0, 1]
print(f"Min-Max: min={scaled_mm.min():.4f}, max={scaled_mm.max():.4f}"
  f"(ожидается: min=0.0, max=1.0)")

# StandardScaler: среднее ≈ 0, std ≈ 1 по каждому столбцу
print(f"Z-Scale: mean={scaled_std.mean(axis=0).round(4)}, "
  f"std={scaled_std.std(axis=0).round(4)}"
  f"(ожидается: mean≈0, std≈1)")

# RobustScaler: медиана ≈ 0 по каждому столбцу
print(f"Robust:median={np.median(scaled_rb, axis=0).round(4)}"
  f"(ожидается: median≈0)")

print('\n--- Анализ ---')
print('Min-Max приводит все значения к диапазону [0, 1], но чувствителен к выбросам:')
print('один экстремальный признак сожмёт остальные значения к нулю.')
print()
print('StandardScaler центрирует данные (среднее=0, std=1), что хорошо для большинства')
print('алгоритмов, но среднее и std смещаются при наличии выбросов.')
print()
print('RobustScaler использует медиану и IQR, поэтому устойчив к выбросам.')
print('Для датасета Iris, где признаки имеют разные масштабы и возможны выбросы,')
print('RobustScaler является наиболее надёжным выбором.')


Min-Max Normalized Data (первые 5 строк):
[[0.22222222 0.625      0.06779661 0.04166667]
 [0.16666667 0.41666667 0.06779661 0.04166667]
 [0.11111111 0.5        0.05084746 0.04166667]
 [0.08333333 0.45833333 0.08474576 0.04166667]
 [0.19444444 0.66666667 0.06779661 0.04166667]]

Z-Normalized Data (первые 5 строк):
[[-0.90068117  1.01900435 -1.34022653 -1.3154443 ]
 [-1.14301691 -0.13197948 -1.34022653 -1.3154443 ]
 [-1.38535265  0.32841405 -1.39706395 -1.3154443 ]
 [-1.50652052  0.09821729 -1.2833891  -1.3154443 ]
 [-1.02184904  1.24920112 -1.34022653 -1.3154443 ]]

Robust Scaled Data (первые 5 строк):
[[-0.53846154  1.         -0.84285714 -0.73333333]
 [-0.69230769  0.         -0.84285714 -0.73333333]
 [-0.84615385  0.4        -0.87142857 -0.73333333]
 [-0.92307692  0.2        -0.81428571 -0.73333333]
 [-0.61538462  1.2        -0.84285714 -0.73333333]]

--- Проверка корректности трансформации ---
Min-Max: min=0.0000, max=1.0000(ожидается: min=0.0, max=1.0)
Z-Scale: mean=[-0. -0. -0. -0

----------------------------------------------------------

# 5: Оптимальное применение метода нормализации
Задание

**Цель:** Оценить эффективность различных классификаторов совместно с разными методами нормализации данных на примере датасета Iris с использованием кросс-валидации.

#### Шаги выполнения задания:

1. **Теоретическая подготовка**
   - Освойте понятия нормализации данных и основные техники нормализации: Min-Max Scaling, Z-нормализация (StandardScaling), Robust Scaling.
   - Изучите три базовых метода классификации: логистическую регрессию, SVM (Support Vector Machine) с линейным ядром и случайный лес (Random Forest).
   - Познакомьтесь с концепцией кросс-валидации и оценкой точности моделей.

2. **Подготовьте окружение**
   - Установите необходимые библиотеки Python, такие как `pandas`, `numpy`, `matplotlib`, `seaborn` и `scikit-learn`.

3. **Получите и подготовьте данные**
   - Загрузите стандартный датасет Iris из библиотеки Scikit-Learn.
   - Разделите признаки и целевую переменную, используя соответствующие структуры данных.

4. **Методика эксперимента**
   - Определите классификационные модели, которые будут использованы в эксперименте:
     - Логистическая регрессия (`LogisticRegression`)
     - Линейный SVM (`SVC(kernel='linear')`)
     - Случайный лес (`RandomForestClassifier(n_estimators=100)`).
   - Выберите три метода нормализации данных:
     - Min-Max normalization (`MinMaxScaler`)
     - Z-нормализация (`StandardScaler`)
     - Робастная нормализация (`RobustScaler`).

5. **Эксперимент**
   - Для каждой комбинации классификатора и метода нормализации выполните следующие шаги:
     - Нормализуйте признаки с помощью выбранного метода.
     - Примените кросс-валидацию с пятью фолдами (`cv=5`) для оценки средней точности классификации.
     - Запишите среднюю точность для каждой пары "метод нормализации + алгоритм классификации".

6. **Интерпретация результатов**
   - Проанализируйте полученные результаты: какая комбинация метода нормализации и классификатора показала наилучшую производительность?
   - Какие факторы могли повлиять на различия в результатах?
   - Предложите рекомендации относительно выбора метода нормализации и классификатора для новых наборов данных.

7. **Доклад и отчет**
   - Подготовьте отчёт, включающий вводную часть, описание методик и экспериментов, анализ полученных результатов и выводы.


In [26]:
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier

# Загружаем датасет Iris и разделяем на признаки (X) и целевую переменную (y)
X, y = load_iris(return_X_y=True)

# Классификационные модели
classifiers = {
    'Logistic Regression': LogisticRegression(max_iter=10000),
    'SVM':                 SVC(kernel='linear'),
    'Random Forest':       RandomForestClassifier(n_estimators=100, random_state=42),
}

# Методы нормализации данных
scalers = {
    'Min-Max': MinMaxScaler(),
    'Z-Scale': StandardScaler(),
    'Robust':  RobustScaler(),
}

# Результаты собираем в список и выводим таблицей
results = []
for clf_name, classifier in classifiers.items():
    for scale_name, scaler in scalers.items():
        scaled_data = scaler.fit_transform(X)
        # Кросс-валидация с 5 фолдами
        scores = cross_val_score(classifier, scaled_data, y, cv=5)
        results.append({
            'Классификатор':    clf_name,
            'Нормализация':     scale_name,
            'Средняя точность': round(scores.mean(), 3),
        })

results_df = pd.DataFrame(results)
print(results_df.to_string(index=False))

# Интерпретация результатов
best_score = results_df['Средняя точность'].max()
best_rows = results_df[results_df['Средняя точность'] == best_score]

print(f"\n--- Интерпретация результатов ---")
for clf_name in results_df['Классификатор'].unique():
  clf_df = results_df[results_df['Классификатор'] == clf_name]
  best_score = clf_df['Средняя точность'].max()
  best_scalers = clf_df[clf_df['Средняя точность'] == best_score]['Нормализация'].tolist()
  print(f"{clf_name}: лучшая нормализация — {', '.join(best_scalers)} "
        f"(точность: {best_score})")

print()
print("Факторы, влияющие на различия результатов:")
print("- SVM и LogisticRegression чувствительны к масштабу признаков, поэтому")
print("  нормализация существенно влияет на их точность.")
print("- Random Forest основан на деревьях решений и инвариантен к масштабу,")
print("  поэтому результаты слабо зависят от метода нормализации.")
print()
print("Рекомендации:")
print("- Для линейных моделей (LR, SVM) предпочтительны Z-Scale или Min-Max.")
print("- При наличии выбросов в данных лучше использовать RobustScaler.")
print("- Random Forest можно применять без нормализации, результат будет схожим.")

      Классификатор Нормализация  Средняя точность
Logistic Regression      Min-Max             0.927
Logistic Regression      Z-Scale             0.960
Logistic Regression       Robust             0.940
                SVM      Min-Max             0.967
                SVM      Z-Scale             0.967
                SVM       Robust             0.960
      Random Forest      Min-Max             0.967
      Random Forest      Z-Scale             0.967
      Random Forest       Robust             0.967

--- Интерпретация результатов ---
Logistic Regression: лучшая нормализация — Z-Scale (точность: 0.96)
SVM: лучшая нормализация — Min-Max, Z-Scale (точность: 0.967)
Random Forest: лучшая нормализация — Min-Max, Z-Scale, Robust (точность: 0.967)

Факторы, влияющие на различия результатов:
- SVM и LogisticRegression чувствительны к масштабу признаков, поэтому
  нормализация существенно влияет на их точность.
- Random Forest основан на деревьях решений и инвариантен к масштабу,
  поэтому 

## Отчёт

### Введение
В данном эксперименте исследуется влияние метода нормализации данных на точность
классификации. Для оценки используется датасет Iris — классический набор данных,
содержащий 150 образцов трёх видов ирисов с четырьмя признаками каждый.

### Описание методик
Рассматриваются три метода нормализации:
- Min-Max: приводит значения к диапазону [0, 1]
- Z-Scale (StandardScaler): центрирует данные (среднее=0, std=1)
- Robust (RobustScaler): использует медиану и IQR, устойчив к выбросам

Рассматриваются три классификатора:
- Logistic Regression — линейная модель
- SVM с линейным ядром — линейная модель, чувствительная к масштабу
- Random Forest — ансамблевая модель на основе деревьев решений

Оценка производится методом кросс-валидации с 5 фолдами.

### Результаты эксперимента
- Logistic Regression показала лучший результат с Z-Scale (0.960)
- SVM показала лучший результат с Min-Max и Z-Scale (0.967)
- Random Forest показала одинаковый результат со всеми методами (0.967)

### Анализ и выводы
Logistic Regression наиболее чувствительна к выбору нормализации: разброс
результатов составил 0.033 (от 0.927 до 0.960). Z-Scale оказался оптимальным,
так как центрирование данных улучшает сходимость линейных моделей.

SVM также чувствителен к масштабу, однако показал высокую точность с Min-Max
и Z-Scale. Robust дал чуть меньший результат (0.960).

Random Forest инвариантен к масштабу признаков — все три метода нормализации
дали одинаковый результат (0.967), что подтверждает теоретические свойства
алгоритма на основе деревьев решений.

Рекомендация: для линейных моделей предпочтителен Z-Scale или Min-Max.
При наличии выбросов в данных следует использовать RobustScaler.
Для Random Forest выбор метода нормализации не имеет значения.